src/nvtxw.rs

```diff
@@ -0,0 +1,427 @@
+use std::collections::BTreeMap;
+use std::ffi::CString;
+use std::ffi::{c_char, c_void};
+use std::io;
+use std::iter::zip;
+use std::mem::size_of;
+use std::path::PathBuf;
+use std::ptr::{null, null_mut};
+
+use nvtxw::nvtxw;
+
+use crate::data::{DataSourceInfo, EntryID, EntryIndex, EntryInfo, SlotMetaTile, SlotTile, TileID};
+use crate::deferred_data::{CountingDeferredDataSource, DeferredDataSource};
+
+const LEGION_DOMAIN_NAME: &str = "Legion";
+
+pub struct NVTXW<T: DeferredDataSource> {
+    data_source: CountingDeferredDataSource<T>,
+    output: Option<PathBuf>,
+    merge: Option<PathBuf>,
```

Comment:
See my note here and apply the same thing:

https://gitlab.com/StanfordLegion/legion/-/merge_requests/1001#note_2076285077

In general, it's considered a best practice not to pass around concrete path types, as `AsRef<Path>` gives you more flexibility in passing around e.g. `OsString` without requiring any conversion.

Reply:
I've tried implementing this, but am getting errors.

```rs

pub fn session_begin_simple<P1, P2>(
    interface: InterfaceHandle,
    output: Option<P1>,
    merge: Option<P2>,
) -> Result<SessionHandle, ResultCode>
where
    P1: AsRef<Path>,
    P2: AsRef<Path>,
{
    let session_begin = interface
        .SessionBegin
        .ok_or(NVTXW3_RESULT_INVALID_ARGUMENT)?;

    let c_output = output
        .as_ref()
        .and_then(|x| {
            let p = Path::new(x);
            // ...
```

```
146 |             let p = Path::new(x);
    |                     --------- ^ the trait `AsRef<OsStr>` is not implemented for `P1`
    |                     |
    |                     required by a bound introduced by this call
    |
```

Reply:
You'd need to do something like: 

```rust
let output = output.map(AsRef::as_ref);
let merge = merge.map(AsRef::as_ref);
```

And then you'll have them both as `Option<Path>` for the rest of the function.

Reply:
```
error[E0631]: type mismatch in function arguments
   --> src/nvtxw.rs:144:29
    |
144 |     let output = output.map(AsRef::as_ref);
    |                         --- ^^^^^^^^^^^^^
    |                         |   |
    |                         |   expected due to this
    |                         |   found signature defined here
    |                         required by a bound introduced by this call
    |
    = note: expected function signature `fn(P1) -> _`
               found function signature `fn(&_) -> _`
note: required by a bound in `Option::<T>::map`
   --> /rustc/3f5fd8dd41153bc5fdca9427e9e05be2c767ba23/library/core/src/option.rs:1070:5
help: consider wrapping the function in a closure
    |
144 |     let output = output.map(|arg0: P1| AsRef::as_ref(&arg0));
    |                             ++++++++++              +++++++

```

Reply:
I've opted for making these `OsString`s instead.